### PR TITLE
Implemented browsing in BrowseController under /browse

### PR DESCRIPTION
**Pull request by sanket407**
Implemented browsing under /browse.
All folders are redirected to /browse again from browsing.
All files are redirected to /storages for download.
Note: File/Artifact download not supported for Nuget currently as NugetArtifactController does not support direct file download currently.
Test this with accept-text/html

### CODE REVIEW

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java [hunk 1]
```diff
@@ -63,13 +71,11 @@
     public ResponseEntity<String> browse()
     {
         try
-        {
-            Map<String,SortedSet<String>> storages = new HashMap<>();
-            storages.put("storages", new TreeSet<String>(configurationManager.getConfiguration()
-                    .getStorages()
-                    .keySet()));
+        {   
+            Path dirPath = getStoragesDir();
```
Comment (sbespalov) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: In the original implementation storages was fetched from configuration. 
@sanket407 have you approved with @carlspring that storages will be treated as folders now?

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java [hunk 2]
```diff
@@ -97,10 +103,10 @@
                                      .body("{ 'error': 'The requested storage was not found.' }");
             }
             
-            Map<String,SortedSet<String>> repos = new HashMap<>();
-            repos.put("repositories", new TreeSet<String>(storage.getRepositories().keySet()));
+            Path dirPath = Paths.get(storage.getBasedir());
```
Comment (sbespalov) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: same as above.
I think that we should fetch storages and repositories form configuration.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java [hunk 3]
```diff
@@ -145,26 +151,17 @@
             
             String matchedPattern = (String) request.getAttribute(HandlerMapping.BEST_MATCHING_PATTERN_ATTRIBUTE);
             String subPath = new AntPathMatcher().extractPathWithinPattern(matchedPattern, request.getRequestURI());
```
Comment (sbespalov) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: take a look how `path` variable implemented in `MavenArtifactController`. You can try to use `{storageId}/{repositoryId}/{path:.+}` to expose it. I think this can suit well here.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java [hunk 4]
```diff
@@ -145,26 +151,17 @@
...
-            Path baseDir = Paths.get(repository.getBasedir().toString(), subPath);
+            Path baseDir = Paths.get(repository.getBasedir().toString(), subPath.replaceAll("/", Matcher.quoteReplacement(File.separator)));
```
Comment (sbespalov) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: you should resolve paths with `RepositoryProvider.resolvePath` method

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java [hunk 5]
```diff
@@ -174,4 +171,277 @@
...
+           
+            for (FileContent fileContent : content.getDirectories())
+            {
+                appendFile(sb, fileContent, requestUri, true, url);
+            }
+                        
+            if(url == null && !content.getFiles().isEmpty())
+            {
+                logger.debug("Files are present outside a repository");
+                response.setStatus(HttpStatus.INTERNAL_SERVER_ERROR.value());
```
Comment (sbespalov) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: can you please follow [How-To-Implement-Spring-Controllers](https://github.com/strongbox/strongbox/wiki/How-To-Implement-Spring-Controllers) guide

**Comment by sbespalov** [comment]
@sanket407 can you please point me where `ArtifactResolutionService.resolveArtifactResource` method used?

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java (hunk 1 above)
Comment (sanket407) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: @sbespalov I had implemented them similarly in BrowseStoragesController. There it was approved. @carlspring What do you suggest ?

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/DirectoryContent.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        this.directories = new ArrayList<FileContent>();
+        this.files = new ArrayList<FileContent>();
+               
+        setDirectories(dirPath);
+        setFiles(dirPath);
+    }
+       
+    public void setDirectories(Path dirPath) throws IOException
+    {   
+        Files.list(dirPath)
```
Comment (sbespalov) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/DirectoryContent.java]: I think this logic should be implemented in some service instead of directly here.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/FileContent.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+{
+
+    private String name;
+
+    private String size;
+
+    private String lastModified;
+    
+    @JsonCreator
+    public FileContent( @JsonProperty("name") String name,
```
Comment (sbespalov) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/FileContent.java]: why you need this `@JsonProperty` here?

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/FileContent.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+                        @JsonProperty("lastModified") String lastModified)
+    {
+        this.name = name;
+        this.size = size;
+        this.lastModified = lastModified;
+    }
+    
+    public FileContent(Path p)
+    {   
+        if(p == null)
```
Comment (sbespalov) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/FileContent.java]: same as above in `DirectoryContent`
you should populate your POJOs in service.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java [hunk 9]
```diff
@@ -174,4 +171,277 @@
...
+        Path baseDir = Paths.get(repository.getBasedir().toString(), subPath.replaceAll("/", Matcher.quoteReplacement(File.separator)));
+                
+        logger.debug("Requested browsing for {}/{}/{} ", storageId, repositoryId, subPath);
+        
+        if (repository.allowsDirectoryBrowsing() && probeForDirectoryListing(repository, subPath))
+        {
+            try
+            {
+                DirectoryContent content = new DirectoryContent(baseDir);
+                URL url = artifactResolutionService.resolveResource(storageId, repositoryId, subPath);
```
Comment (sanket407) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: @sbespalov  1.  Here I retrieved the URL and used it later in appendFile !

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java [hunk 10]
```diff
@@ -174,4 +171,277 @@
...
+        sb.append("<tr>");
+
+        if(isDirectory)
+        {
+            sb.append("<td><a href='" + URLEncoder.encode(name, "UTF-8") + 
+                      "/'>" + name + "/" + "</a></td>");
+        }
+        else
+        {  
+            String storagesUrl = url.toString() + "/" + URLEncoder.encode(name, "UTF-8");
```
Comment (sanket407) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: @sbespalov 2. Used the URL here for getting the URL pointing to /storages

Comment (sbespalov) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: I think `artefactUrl` will be more meaningfully instead of `storagesUrl`.
Am I understand correctly that this is URL to download artefact?

Comment (sanket407) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: Yes you are right!

Comment (sbespalov) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: also, why you concatenate `name` to URL?
this URL assumed to be ready to use without modifications

Comment (sanket407) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: Because URL is URL for current folder browse request  which only points to parent folder of files. So I resolve path for parent folder and append filename.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java (hunk 1 above)
Comment (carlspring) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: I believe @sbespalov has a point here. The thing is this: sure, one would normally think that clicking ".." should take you one level up, but then again... what happens if you have some sensitive directories which you've put there, (as an admin), directly on the file-system? For example, you've disabled (commented out the configuration for this in the `strongbox.xml`) some storage for the moment, or have some back up there on the file system... Since it's disabled, people shouldn't be able to browse it.
But... then you would have to the the same for repositories as well...

So, in conclusion, I believe @sbespalov is right here and we should do it as it was (and also apply this change to repositories).

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java (hunk 2 above)
Comment (carlspring) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: +1

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java (hunk 4 above)
Comment (carlspring) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: +1

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java (hunk 5 above)
Comment (carlspring) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: +1

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/DirectoryContent.java (hunk 6 above)
Comment (carlspring) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/DirectoryContent.java]: +1

File: strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/BrowseControllerTest.java [hunk 11]
```diff
@@ -56,13 +76,16 @@ public void testGetStorages()
...
-        assertFalse("Returned storage size does not match", returnedMap.get("storages").isEmpty());
+        DirectoryContent returned = new ObjectMapper()
+                .readValue(storages, DirectoryContent.class);
+              
+        assertNotNull("Failed to get storage list!", returned);
+        assertNotNull("Failed to get storage list!", returned.getDirectories());
+        assertFalse("Returned storage size does not match", returned.getDirectories().isEmpty());
+        
+        for(int i=0; i<returned.getDirectories().size(); i++)
+            logger.debug(returned.getDirectories().get(i).getName());
```
Comment (carlspring) [inline strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/BrowseControllerTest.java]: Please, always use braces in cycles and conditions.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/FileContent.java (hunk 7 above)
Comment (sanket407) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/FileContent.java]: https://stackoverflow.com/questions/21920367/why-when-a-constructor-is-annotated-with-jsoncreator-its-arguments-must-be-ann

P.S. : This is my first time using JsonCreator.

Comment (sbespalov) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/FileContent.java]: ok, why do you need `JsonCreator` here? :)

Comment (sanket407) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/FileContent.java]: Getting foll. error without it : 
com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `org.carlspring.strongbox.domain.DirectoryContent` (no Creators, like default construct, exist): cannot deserialize from Object value (no delegate- or property-based Creator)

I think since we do not have a kind of default constructor for our POJO, the object mapper cannot create the JSON object from the String as it does not know how to. So this constructor points  the deserializer to this JsonCreator for making JSON object

Comment (sbespalov) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/FileContent.java]: > P.S.: this error occurred in Junit Test where I need to deserialize the string to object.

oh, now it's clear :)

Comment (sbespalov) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/FileContent.java]: but if you will create noags constructor and add setters then you can remove this annotations from constructor and it will look much more clearer

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java (hunk 10 above)
Comment (sbespalov) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: the main thing of `artifactResolutionService.resolveResource` method is to avoid additional modifications for URL you get from it, so I think in case of artifact files you need to construct URL just with this method without these modifications

**Comment by carlspring** [comment]
@sanket407 ,

What is the state of this task?

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java [hunk 12]
```diff
@@ -63,13 +79,21 @@
...
+        {               
+            Set<Entry<String, Storage>> storages = configurationManager.getConfiguration().getStorages().entrySet();
+                                                    
+            List<Path> storagePaths = new ArrayList<Path>();
+            
+            for(Entry<String, Storage> entry : storages)
+            {
+                Storage storage = entry.getValue();
+                String storageBaseDir = storage.getBasedir();
+                storagePaths.add(Paths.get(storageBaseDir));
```
Comment (sbespalov) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: why you are using `Paths` here again? we agreed that storages and repositories should be fetched only from configuration.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java [hunk 13]
```diff
@@ -97,10 +121,20 @@
...
+            Set<Entry<String, Repository>> repositories = storage.getRepositories().entrySet();
+            
+            List<Path> repositoryPaths = new ArrayList<Path>();
             
-            return ResponseEntity.ok(objectMapper.writer().writeValueAsString(repos));
+            for(Entry<String, Repository> entry : repositories)
+            {
+                Repository repository = entry.getValue();
+                String repositoryBaseDir = repository.getBasedir();
+                repositoryPaths.add(Paths.get(repositoryBaseDir));
```
Comment (sbespalov) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: you shouldn't use `Paths`, use `RepositoryProvider` instead.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/DirectoryContent.java [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.List;
+import javax.inject.Inject;
+import org.carlspring.strongbox.services.FileContentFetcher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DirectoryContent
+{       
+    @Inject
+    FileContentFetcher fileContentFetcher;
```
Comment (sbespalov) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/DirectoryContent.java]: I think you shouldn't inject anything here

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java (hunk 12 above)
Comment (sanket407) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: Well I have fetched storages and repositories from configuration . Then only used Paths to get the lastmodified and size from file system for ONLY THOSE repositories and storages.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/DirectoryContent.java (hunk 14 above)
Comment (sanket407) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/DirectoryContent.java]: So I should create NEW instance of FileContentFetcher ?

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java (hunk 12 above)
Comment (sbespalov) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: as I see you fetch storage content from FS here, and this is wrong
you should list repositories which defined in configuration instead

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/DirectoryContent.java (hunk 14 above)
Comment (sbespalov) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/DirectoryContent.java]: you shouldn't have `FileContentFetcher` here at all

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java (hunk 12 above)
Comment (sbespalov) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: oh, I see now what you mean 
the thing is that `fetchDirectoryContent` method name confusing a little :)

Comment (sbespalov) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: also, @carlspring do we need size attribute calculated for directories? in general do we need any attributes for the directories?

Comment (sanket407) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: @sbespalov  @carlspring  Yes I was going to mention the  same for 'size' attribute . The way I have seen on some similar listings online, they usually have a dash ( - ) in place of size for directories.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/DirectoryContent.java (hunk 14 above)
Comment (sanket407) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/domain/DirectoryContent.java]: Got you now. I thought this was the DirectoryContentFetcher .

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java (hunk 12 above)
Comment (carlspring) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: @sanket407 : Yes, this seems right. You can ahead and change it.

Comment (sanket407) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: @carlspring  Done !

Comment (carlspring) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/BrowseController.java]: @sbespalov : Sorry, I've just realized that this is actually a very bad idea. Every time you browse, the size of the directories will have to be re-calculated. The more your repositories grow, the slower the browsing will be.
@sanket407 : I'm very sorry for realizing this so late, but would you mind reverting this particular change?

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/services/FileContentFetcher.java [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+        
+        FileContent fileContent = new FileContent();
+        
+        String name = fileName.toString();
+        String size = FileUtils.byteCountToDisplaySize(path.toFile().length());
+        String lastModified = new SimpleDateFormat("dd-MM-yyyy HH-mm-ss")
+                                .format(new Date(path.toFile().lastModified()));
+        
+        fileContent.setName(name);
+        if(Files.isDirectory(path))
```
Comment (sanket407) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/services/FileContentFetcher.java]: @carlspring Here it is !
